ntoskrnl/mm/mmfault.c

```diff
@@ -239,7 +239,10 @@ MmAccessFault(IN ULONG FaultCode,
     if (MmGetKernelAddressSpace())
     {
         /* Check if this is an ARM3 memory area */
+        ASSERT(KeGetCurrentIrql() < DISPATCH_LEVEL);
+        MmLockAddressSpace(MmGetKernelAddressSpace());
```

Comment:
Can we have an ASSRT in MmLocateMemoryAreaByAddress?
Something like
`ASSERT(PsGetCurrentProcess()->AddressCreationLock.Owner != KeGetCurrentThread());`

Reply:
@tkreuzer, Sorry, I do not understand your question. Are you asking me to add this statement into the code at MmLocateMemoryAreaByAddress? Or was this question for another developer? I need help here. Thanks.

Reply:
The first. I formulated it as a question, because I don't know if that will actually work. But please try that.

Reply:
I added what you see in the screen capture below. On boot there is a constant stream of this assert being hit. The debug log is attached for your information as well. I was not able to get past the first black screen. Let me know if you need anything else. Thanks.
![marea patch-01](https://github.com/reactos/reactos/assets/32620577/97fe8ccc-84d8-4716-bbc5-7616882d827f)
[PR-5658-test.txt](https://github.com/reactos/reactos/files/12682433/PR-5658-test.txt)

Reply:
@Doug-Lyons When you get this assertion hit, could you display in the windbg the value of Process, Process->AddressCreationLock (and the Owner field as well) ?
```
kd> ??Process
...
kd> ??Process->AddressCreationLock
...
```
etc.

Reply:
@HBelusca, See Below.
[PR-5658-test-02.txt](https://github.com/reactos/reactos/files/12683691/PR-5658-test-02.txt)

Reply:
OK so we have for example,
```
kd> ?? Process
struct _EPROCESS * 0xb13f9bb0
[...]
   +0x094 UniqueProcessId  : 0x00000004 Void
[...]
   +0x164 ImageFileName    : [16]  "System"
[...]
kd> ?? Process->AddressCreationLock
struct _KGUARDED_MUTEX
   +0x000 Count            : 0n1
   +0x004 Owner            : (null) 
   +0x008 Contention       : 0
   +0x00c Gate             : _KGATE
   +0x01c KernelApcDisable : 0n0
   +0x01e SpecialApcDisable : 0n-1
   +0x01c CombinedApcDisable : 0xffff0000
kd> ?? Process->AddressCreationLock.Owner
struct _KTHREAD * 0x00000000
```

Reply:
@HBelusca, after seeing the the .Owner was NULL, I modified the IF to what is shown in the screen shot below and still had the assert triggered as shown in the debug log below.
![marea patch-02](https://github.com/reactos/reactos/assets/32620577/eed69c4c-2c55-47d0-8393-e02d08e87d47)
[PR-5658-test-03.txt](https://github.com/reactos/reactos/files/12683906/PR-5658-test-03.txt)

Reply:
For the case of `Process == NULL`, you can add `ASSERT(PsIdleProcess->AddressCreationLock.Owner == KeGetCurrentThread());`

Reply:
I had a look and found missing locks in `MiProtectVirtualMemory` and `MmAccessFault`. So you need to fix those 2 instance, too.

Reply:
@tkreuzer, Please review these changes and see if they are what you had in mind. Thanks lots.

Reply:
Yes looks good. Do you still get any assertion failures?

Reply:
I have only tested booting and running RAPPS to download Total Commander and this happens without triggering any asserts. I should probably do a full compile of ReactOS on ReactOS and then a set of regression tests to say for sure that there are no asserts. I will try and do these things later today. Thanks for your help and attention to detail with this.

Reply:
Can you also add the
```
    else
        ASSERT(PsIdleProcess->AddressCreationLock.Owner == KeGetCurrentThread());
```

Reply:
@tkreuzer. I am waiting to make this commit until we can get the KVM_x64 bot to run. Thanks.

Reply:
Sorry, @tkreuzer after looking more closely at your request, I am not sure where to add this "else...". Can you be more specific? I have removed the comment slashes now. Thanks.

Reply:
Should be fixed now. Thanks.